jems-testing/src/test/java/org/dmfs/jems/pair/hamcrest/PairMatcherTest.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2017 dmfs GmbH
+ *
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.dmfs.jems.pair.hamcrest;
+
+import org.dmfs.jems.pair.elementary.ValuePair;
+import org.dmfs.testutils.tools.AbstractValueObject;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+import static org.dmfs.jems.pair.hamcrest.PairMatcher.pair;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.sameInstance;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+
+/**
+ * Unit test for {@link PairMatcher}.
+ *
+ * @author Gabor Keszthelyi
+ */
+public final class PairMatcherTest
+{
+
+    @Rule
+    public ExpectedException exception = ExpectedException.none();
+
+
+    @Test
+    public void test_isPairWithNoMatcher_defaultsToEqualTo()
+    {
+        assertThat(new ValuePair<>(new LeftValueObject(1), new RightValueObject(5)),
+                PairMatcher.pair(new LeftValueObject(1), new RightValueObject(5)));
+    }
+
+
+    @Test
+    public void test_isPairWithMatcher()
+    {
+        Left left = new Left();
+        Right right = new Right();
+
+        assertThat(new ValuePair<>(left, right),
+                pair(sameInstance(left), sameInstance(right)));
+
+        assertThat(new ValuePair<>(left, new RightValueObject(3)),
+                pair(sameInstance(left), equalTo(new RightValueObject(3))));
+
+        assertThat(new ValuePair<>(new LeftValueObject(2), right),
```

Comment:
This looks like overtesting to me. You essentially test the exact same code 3 times. All three test cases are about passing `Matcher`s to `pair(…)`. How could the 2nd or 3rd test fail if the first one passes (assuming that `sameInstance` and `equalTo` are already well tested)?
This single line should give the same amount of certainty:
```java
 assertThat(new ValuePair<>(new String("a"), new Integer(1)),
                pair(equalTo(new String("a")), equalTo(new Integer(1))));
```

That being said, the test is broken. You write it as if `ValuePair` is under test.

The test should actually look like
```java
assertThat(
        pair(equalTo(new String("a")), equalTo(new Integer(1)))
            .matches(new ValuePair<>(new String("a"), new Integer(1))),
         is(true)
);
assertThat(
        pair(equalTo(new String("b")), equalTo(new Integer(1)))
            .matches(new ValuePair<>(new String("a"), new Integer(1))),
         is(false)
);
assertThat(
        pair(equalTo(new String("a")), equalTo(new Integer(2)))
            .matches(new ValuePair<>(new String("a"), new Integer(1))),
         is(false)
);
assertThat(
        pair(equalTo(new String("b")), equalTo(new Integer(2)))
            .matches(new ValuePair<>(new String("a"), new Integer(1))),
         is(false)
);
```
The same applies to the other tests. Also using `ValuePair` here means we can not test `ValuePair` with this Matcher, which is fine but you could consider using a mock here. Not that it really matters, though.

Reply:
Right, the test should be about testing `matches`, I've changed it. Also using mock now instead of `ValuePair`.
Ready for review again.